### ISSUE TITLE
Step tuner priorities have no effect

### ISSUE THREAD

**Issue by DaoWen** [icnc/icnc#33]
I was playing with the step priority tuning support in icnc, and I thought it was strange that I always got basically the same performance, even when I set a priority which should have really hurt the performance in my application.

I put together a very minimal test to see if priorities were doing anything. I figured if I created 10 steps and ran them on a single thread, then they would execute in order of their priority. However, my observation is that they just execute in FIFO order:
```
[CnC] Bind thread 0 to processor 0
[CnC] Using FIFO_SINGLE scheduler [PRIORITY ON] [PINNING ON]
Step priority is 0
Step priority is 3
Step priority is 6
Step priority is 9
Step priority is 2
Step priority is 5
Step priority is 8
Step priority is 1
Step priority is 4
Step priority is 7
Executing step 0
Executing step 3
Executing step 6
Executing step 9
Executing step 2
Executing step 5
Executing step 8
Executing step 1
Executing step 4
Executing step 7
Steps created( 10 )
Steps scheduled( 10 ) inflight( 0, 0 )
Steps requeued( 0 ) resumed( 0 )
```

This simple test app is added in PR #32. See the readme included with the example for instructions on how to build and run the app.

I'm using the latest binary distribution (with TBB bundled) on a FC19 x86_64 box.

**Comment by fschlimb** [comment]
Thanks. Apparently this is a problem when using a single thread only. It works fine with multiple threads. I'll look into this more.

**Comment by DaoWen** [comment]
The tasks still print in the same order for me, even when running with 2 or 3 threads (with the delay slightly altered to stagger the output).

**Comment by DaoWen** [comment]
I've confirmed that this is only a problem for a single thread. This is because the `steal` behavior is disabled when only using one thread:

https://github.com/icnc/icnc/blob/62aa3ca/src/tbb_concurrent_queue_scheduler.cpp#L188

When `steal` is false, the scheduler uses `m_gQueue` (the "global" queue?) for all push/pop operations, which has a fixed type of `tbb::concurrent_bounded_queue` (unlike the parameterized type used for `m_queues`):

https://github.com/icnc/icnc/blob/62aa3ca/src/tbb_concurrent_queue_scheduler.h#L86

Basically, running with a single thread disables the "work stealing" behavior, which in turn causes the scheduler to use a task queue with no priority support.

**Comment by fschlimb** [comment]
good catch!
So I guess I will have to see if I can use the priority queue as m_gQueue and then also fix the queue creation for these.

**Comment by DaoWen** [comment]
> So I guess I will have to see if I can use the priority queue as m_gQueue and then also fix the queue creation for these.

That's what I tried to do initially with my patch—but since the `tbb::concurrent_bounded_queue` doesn't support the blocking `push` and `pop` operations used by `m_gQueue`, there wasn't a simple way to get it working. However, since you're more familiar with TBB, you can probably put together a wrapper to add those or something.

**Comment by fschlimb** [comment]
Oh, now I remember...yes, that's an issue.
